src/lib/Ouroboros/Byron/Proxy/Main.hs

```diff
@@ -133,42 +142,178 @@ data ByronProxy = ByronProxy
-  { -- This is needed by the old relay system, so it can make the INV
-    -- announcements after a DATA messages has been processed.
-    toKey = pure . tagWith ptag . keyFromValue
-    -- Handle an INV: True if we don't have it, False if we do.
-  , handleInv = \k -> fmap (maybe True (const False)) $ atomically $
-      Pool.lookup (untag k) pool
-    -- Handle a REQ: Nothing if we don't have it, Just if we do.
-  , handleReq = \k -> atomically $ Pool.lookup (untag k) pool
-    -- Handle a DATA: put it into the pool, process it, and give False to mean it
-    -- should _not_ be relayed. For this Byron proxy, we'll never relay data
-    -- received from Byron to another Byron node.
-  , handleData = \v -> do
-      atomically $ Pool.insert (keyFromValue v) v pool
-      process v
-      pure False
+  -> KeyVal (Tagged tag k) v m
+taggedKeyValNoOp ptag keyFromValue = KeyVal
+  { toKey      = pure . tagWith ptag . keyFromValue
+  -- We don't want it.
+  , handleInv  = const (pure False)
+  -- We don't have it.
+  , handleReq  = const (pure Nothing)
+  -- We don't care about it and will not relay it.
+  , handleData = const (pure False)
   }
 
-type TxPool = Pool TxId TxMsgContents
-type UpProposalPool = Pool UpId (UpdateProposal, [UpdateVote])
-type UpVotePool = Pool VoteId UpdateVote
-type SscCommitmentPool = Pool StakeholderId MCCommitment
-type SscOpeningPool = Pool StakeholderId MCOpening
-type SscSharesPool = Pool StakeholderId MCShares
-type SscVssCertPool = Pool StakeholderId MCVssCertificate
+-- | Make a `KeyVal` for transactions backed by a `Mempool`.
+-- The `TVar` is expected to be kept somewhat in sync with the actual contents
+-- of the mempool, which works on monotonically-increasing ticket numbers and
+-- has no index based on hash (which Byron requires).
+txKeyValFromMempool
+  :: forall cfg .
+     ( ApplyTx (Block cfg) )
+  => Mempool IO (Block cfg) TicketNo
+  -> TVar (Map TxId TicketNo)
+  -> KeyVal (Tagged TxMsgContents TxId) TxMsgContents IO
+txKeyValFromMempool mempool txTicketMapVar = KeyVal
+    { toKey = pure . tagWith ptag . keyFromValue
+    -- Peer offers this transaction. Give True if we don't have it (we want it)
+    -- and False otherwise.
+    , handleInv = \txId -> atomically $ do
+        snapshot <- Tx.Out.mempoolGetSnapshot reader
+        ticketMap <- readTVar txTicketMapVar
+        let outcome = do
+              ticketNo <- Map.lookup (untag txId) ticketMap
+              Tx.Out.mempoolLookupTx snapshot ticketNo
+        pure $ maybe True (const False) outcome
+    -- Peer requests this transaction. Give Nothing if we don't have it, and
+    -- Just if we do.
+    , handleReq = \txId -> atomically $ do
+        snapshot <- Tx.Out.mempoolGetSnapshot reader
+        ticketMap <- readTVar txTicketMapVar
+        let outcome = do
+              ticketNo <- Map.lookup (untag txId) ticketMap
+              Tx.Out.mempoolLookupTx snapshot ticketNo
+        pure $ fmap toByronTxMsgContents outcome
```

Comment:
If somebody asks for it I can factor this out, since it can be used to derive `handleInv`.

Reply:
Probably a good idea.

Reply:
But they are both right next to each other, so if one will need to change it will be easy to spot that the other needs to be updated as well; I'll leave it up to you.